pkg/agent/cniserver/interface_configuration_windows.go

```diff
@@ -504,3 +522,54 @@ func (ic *ifConfigurator) getInterceptedInterfaces(
 func (ic *ifConfigurator) getOVSInterfaceType() int {
 	return internalOVSInterfaceType
 }
+
+func (ic *ifConfigurator) addAsyncFuncRequest(ifName string, myFunc func(ifName string, param interface{}) error, param interface{}) {
+	ic.ifQueue.Add(&asyncIfFunc{ifName: ifName, ifFunc: myFunc, param: param})
```

Comment:
It seems the asyncFuncs don't have much in common. 
And creating several parameter structs and introducing a workqueue seems more complicated than previous code while it doesn't make it more efficient as all requests will be handled sequentially. Given the new mode, it's likely we will add another `asyncIfFunc` if we need to execute another step after the interface is up in the future, which makes it less efficient.

So I wonder if we could just extract the goroutine in `connectInterfaceToOVSAsync` as a common function and pass callback functions to it. For example:
```
type postInterfaceCreateHook func() error

func (ic *ifConfigurator) addPostInterfaceCreateHook(containerID, endpointName string, containerAccess *containerAccessArbitrator, hook postInterfaceCreateHook) error {
	if containerAccess == nil {
		return fmt.Errorf("container lock cannot be null")
	}
	expectedEp, ok := ic.getEndpoint(endpointName)
	if !ok {
		return fmt.Errorf("failed to find HNSEndpoint %s", endpointName)
	}
        ...
        hook()
}

func (pc *podConfigurator) connectInterfaceToOVSAsync(ifConfig *interfacestore.InterfaceConfig, containerAccess *containerAccessArbitrator) error {
	return pc.ifConfigurator.addPostInterfaceCreateHook(ifConfig.ContainerID, ifConfig.InterfaceName, containerAccess, func() error {
		return pc.connectInterfaceToOVSCommon(ifConfig.InterfaceName, ifConfig)
	})
}

// configureContainerLink creates a HNSEndpoint for the container using the IPAM result, and then attach it on the container interface.
func (ic *ifConfigurator) configureContainerLink(...) {
        ...
	ifaceName := fmt.Sprintf("vEthernet (%s)", endpoint.Name)
	return ic.addPostInterfaceCreateHook(containerID, epName, containerAccess, func() error {
		return util.SetInterfaceMTU(ifaceName, mtu)
	})
}
```

Reply:
It is a great idea to leverage the callback to unify the configuration functions, I would change it.

And I have concern on CPU increasement caused by the number of goroutines in scale setup, if we create a new goroutine for each configuration (e.g., OVS port attachment, and MTU configuration) to wait for the host interface. For example,  100 Pods are created at the same time using containerd, then we should create 100*2 goroutines to wait the interfaces ready. If we have to add another configuration on the interface in future, then the number of goroutines becomes 100 * 3. @tnqn

Reply:
The operations are not CPU intensive and will be either sleeping or making external call, and a Node can hold at most 100 Pods normally, so even in worse case 200~300 non-CPU-intensive goroutines seem not bad. The memory consumption should be fine too: 2k*300=600k.

Reply:
Got, I would change.